Edit Site: New page based navigation.

epiqueras opened the pull request:
Closes #19257

## Description

This PR adds a new way of navigating in the site editor.

It adds a new page switcher dropdown that lets you select from your list of pages and category pages. The list can be expanded to include other pages in the future. The idea is to have one option for every unique URL path that the site exposes, except for post permalinks. A next step here would be to add a Post section that lets you load the single post page with a specific post context using a nested dropdown with the link UI.

Selecting a page switches the current template to the one resolved for that page and sets the correct root block context, so that post and query blocks work as expected. This flow is shown in the GIF below, where switching to the contact page makes the Post Title block render its title. Category pages also work like this, but the query blocks still need some work to read root context so that part is not testable just yet.

The next step would be to modify the template switcher so that it not only switches the current template but also provides a way to assign it to the current page.

This PR also:

- Fixes some subtle bugs that were introduced by the new API refactors and optimizes template `auto-draft` creation so that they are only recreated when the underlying block template files have changed.
- Adds a fallback `index` template to the demo templates to facilitate testing.

## How has this been tested?

The flow that is shown in the GIF where switching pages in the site editor will update the current template and root block context was tested with a combination of different templates and pages to verify that the correct template is resolved for each page and that the root block context is set correctly.

## Screenshots

![gif](https://user-images.githubusercontent.com/19157096/81984740-1676ed00-95ea-11ea-8277-5a42601796ee.gif)

## Types of Changes

*New Feature:* The site editor now has a page switcher for navigating between pages instead of templates.

## Checklist:

- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

epiqueras committed:
Lib: Add a fallback index demo block template.

epiqueras committed:
Lib: Fix template API refactor bugs and integrate with API for resolving a template for a URL.

epiqueras committed:
Lib: Provide initial page path and context for the site editor.

epiqueras committed:
Edit Site: Implement page switcher.

epiqueras committed:
Lib: Fix more refactor bugs.

epiqueras committed:
Lib: Fix API refactor bugs.

epiqueras committed:
Query Blocks: Add category selector.

epiqueras committed:
Lib: Fix template loader.

vindl reviewed (commented):
Glad to see this coming together! 🚀 

Here are some of observations from my recent testing:

The page dropdown button is not visible on initial load.

<img width="416" alt="Screenshot 2020-05-15 at 20 01 59" src="https://user-images.githubusercontent.com/1182160/82082979-3b9e5500-96e9-11ea-9d92-2cda74295988.png">

Although it does get populated with a page title after clicking on that area and selecting one of the pages. The position of dropdown shifts slightly while toggling pages, depending on the size of the title.

There are also some test failures that seem to be related to this PR.

epiqueras committed:
Lib: Add the global query category IDs to the root block context.

epiqueras committed:
Lib: Support blog front pages in the site editor.

epiqueras committed:
Lib: Add a minimal category demo block template.

epiqueras committed:
Site Editor: Adapt Query Blocks to work with the global query.

epiqueras commented:
>The page dropdown button is not visible on initial load.

This is because you don't have a static front page, and the PR didn't include the "All Posts" page yet. I just added it.

>The position of dropdown shifts slightly while toggling pages, depending on the size of the title.

We need to combine the two dropdowns in another PR.

>There are also some test failures that seem to be related to this PR.

Fixtures, updated.

I also added support for category pages now:

![gif](https://user-images.githubusercontent.com/19157096/82092988-1bea3b00-96af-11ea-9146-8a014064977f.gif)

youknowriad commented on packages/edit-site/src/components/header/index.js:
```diff
@@ -56,6 +67,33 @@ export default function Header( { openEntitiesSavedStates } ) {
 	);
+	const setActivePage = useCallback( async ( newPage ) => {
+		try {
+			const { success, data } = await fetch(
```
Just stumbled upon this and it seems a bit hacky. What does this fetch do? Why not core-data? Why not apiFetch?

youknowriad replied:
Also, it seems to me that all the logic here is better on an async store action. (generator with controls) as otherwise there's a lot of potential for breakage as written (set state on unmounted components).

youknowriad replied:
I guess  the issue is that "settings" is not yet on a store, for me, this is becoming more and more a priority as we add components and share state.

epiqueras replied:
It's not precisely an API endpoint. It's the only straightforward way of triggering Core's template resolution for a URL because calling `parse_request` in WP twice is not trivial.

Yes, I agree that at this point, it makes sense to move settings into a store and have this be an action.

youknowriad replied:
I'm also wondering if we're making things hard for us by trying to act like a full  SPA here where we could instead trigger a reload and append something to the URL? (The dirty detection check could help us to avoid leaving the page with unsaved changes).

epiqueras replied:
We could add deep linking without refreshes too.

I do think that linking to a specific page in the site editor from different places would be nice.

noahtallen replied:
> I'm also wondering if we're making things hard for us by trying to act like a full SPA here where we could instead trigger a reload and append something to the URL?

IMO, it would be nice to avoid hard refreshes. One of the biggest user complaints on WordPress.com is how long the block editor takes to load, which is a combination of loading an iframe and loading the block editor. Destroying the existing JS context and rebuilding it seems to be really inefficient (IMO) when the code can already handle switching pages client side! Since switching pages and templates seems to be a common task, it’d be nice to keep that interaction quick :)
